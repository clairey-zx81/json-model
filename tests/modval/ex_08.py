#! /bin/env python
#
# Model: ./ex_08.model.json
#
# Generated by jmc (JSON Model Compiler) (https://github.com/clairey-zx81/json-model)
#

from typing import Callable
import re as re
import datetime
import urllib.parse

type Jsonable = None|bool|int|float|str|list[Jsonable]|dict[str, Jsonable]
type Path = list[str]
type Report = list[str]|None
type CheckFun = Callable[[Jsonable, str, Report], bool]
type PropMap = dict[str, CheckFun]
type TagMap = dict[None|bool|float|int|str, CheckFun]

# extract type name
def _tname(value: Jsonable) -> str:
    return type(value).__name__

# maybe add message to report
def _rep(msg: str, rep: Report) -> bool:
    rep is None or rep.append(msg)
    return False


def _is_valid_url(value: Jsonable, path: str, rep: Report = None) -> bool:
    if isinstance(value, str):
        try:
            urllib.parse.urlparse(value)
            return True
        except Exception as e:
            rep is None or re.append(f"invalid url at {path}: {value} ({e})")
            return False
    rep is None or rep.append(f"incompatible type for url at {path}: {_tname(value)}")
    return False

# define "$VAL" ($.'$VAL')
def json_model_2(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $.'$VAL'
    result = isinstance(value, bool)
    if not result:
        rep is None or rep.append(f"not a bool at {path} [$.'$VAL']")
    return result

# define "$KEY" ($.'$KEY')
def json_model_3(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $.'$KEY'
    result = _is_valid_url(value, path, rep) or _rep(f"invalid $URL at {path}", rep)
    if not result:
        rep is None or rep.append(f"not an expected $URL at {path} [$.'$KEY']")
    return result


# object $.'$map'
def json_model_4(value: Jsonable, path: str, rep: Report = None) -> bool:
    if not isinstance(value, dict):
        rep is None or rep.append(f"not an object at {path} [$.'$map']")
        return False
    for prop, val in value.items():
        assert isinstance(prop, str)
        lpath = path + "." + prop
        if _is_valid_url(prop, path, rep) or _rep(f"invalid $URL at {path}", rep) or _rep(f"prop {prop} does not match $URL at {path}", rep):  # $URL
            # $.'$map'.'$URL'
            result = json_model_2(val, path, rep)
            if not result:
                rep is None or rep.append(f"not an expected $VAL at {lpath} [$.'$map'.'$URL']")
            if not result:
                return False
        else:  # no catch all
            rep is None or rep.append(f"no other prop expected at {path} [$.'$map']")
            return False
    return True



# define "$EX08" ($.'$EX08')
def json_model_5(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $.'$EX08'
    # $.'$EX08'.'|'.0
    result = json_model_4(value, path, rep)
    if not result:
        rep is None or rep.append(f"not an expected $map at {path} [$.'$EX08'.'|'.0]")
    if not result:
        # $.'$EX08'.'|'.1
        result = _is_valid_url(value, path, rep) or _rep(f"invalid $URL at {path}", rep)
        if not result:
            rep is None or rep.append(f"not an expected $URL at {path} [$.'$EX08'.'|'.1]")
        if not result:
            # $.'$EX08'.'|'.2
            result = json_model_2(value, path, rep)
            if not result:
                rep is None or rep.append(f"not an expected $VAL at {path} [$.'$EX08'.'|'.2]")
    if not result:
        rep is None or rep.append(f"not any model match at {path} [$.'$EX08'.'|']")
    return result

# define "$" ($)
def json_model_1(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $
    result = json_model_5(value, path, rep)
    if not result:
        rep is None or rep.append(f"not an expected $EX08 at {path} [$]")
    return result

# entry function check_model
def check_model(value: Jsonable, path: str = "$", rep: Report = None) -> bool:
    return json_model_1(value, path, rep)


# possibly run as a script: $0 values...
if __name__ == "__main__":
    import json
    import sys
    for fn in sys.argv[1:]:
        try:
            with open(fn) as f:
                value = json.load(f)
            reasons = []
            if check_model(value, "", reasons):
                print(f"{fn}: PASS")
            else:
                print(f"{fn}: FAIL {reasons}")
        except Exception as e:
            print(f"{fn}: ERROR ({e})")

