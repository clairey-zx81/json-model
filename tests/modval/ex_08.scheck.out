ex_08.0.true.json: PASS
ex_08.1.true.json: PASS
ex_08.2.true.json: PASS
ex_08.3.true.json: PASS
ex_08.4.false.json: FAIL ["not an object at $ [$.'$map']", "not an expected $map at $ [$.'$EX08'.'|'.0]", 'incompatible type for url at $: list', 'invalid $URL at $', "not an expected $URL at $ [$.'$EX08'.'|'.1]", "not a bool at $ [$.'$VAL']", "not an expected $VAL at $ [$.'$EX08'.'|'.2]", "not any model match at $ [$.'$EX08'.'|']", 'not an expected $EX08 at $ [$]']
ex_08.5.false.json: FAIL ["not an object at $ [$.'$map']", "not an expected $map at $ [$.'$EX08'.'|'.0]", 'incompatible type for url at $: float', 'invalid $URL at $', "not an expected $URL at $ [$.'$EX08'.'|'.1]", "not a bool at $ [$.'$VAL']", "not an expected $VAL at $ [$.'$EX08'.'|'.2]", "not any model match at $ [$.'$EX08'.'|']", 'not an expected $EX08 at $ [$]']
ex_08.6.false.json: FAIL ["not a bool at $ [$.'$VAL']", "not an expected $VAL at $.foo [$.'$map'.'$URL']", "not an expected $map at $ [$.'$EX08'.'|'.0]", 'incompatible type for url at $: dict', 'invalid $URL at $', "not an expected $URL at $ [$.'$EX08'.'|'.1]", "not a bool at $ [$.'$VAL']", "not an expected $VAL at $ [$.'$EX08'.'|'.2]", "not any model match at $ [$.'$EX08'.'|']", 'not an expected $EX08 at $ [$]']
